zerver/lib/thumbnail.py

```diff
@@ -21,6 +26,71 @@
 # Reject emoji which, after resizing, have stills larger than this
 MAX_EMOJI_GIF_FILE_SIZE_BYTES = 128 * 1024  # 128 kb
 
+T = TypeVar("T", bound="BaseThumbnailFormat")
+
+
+@dataclass(frozen=True, eq=True)
+class BaseThumbnailFormat:
+    extension: str
+    max_width: int
+    max_height: int
+    animated: bool
+
+    @override
+    def __eq__(self, other: object) -> bool:
+        if not isinstance(other, BaseThumbnailFormat):
+            return False
+        return str(self) == str(other)
+
+    @override
+    def __str__(self) -> str:
+        animated = "-anim" if self.animated else ""
+        return f"{self.max_width}x{self.max_height}{animated}.{self.extension}"
+
+    @classmethod
+    def from_string(cls: Type[T], format_string: str) -> Optional[T]:
+        format_parts = re.match(r"(\d+)x(\d+)(-anim)?\.(\w+)$", format_string)
+        if format_parts is None:
+            return None
+
+        return cls(
+            max_width=int(format_parts[1]),
+            max_height=int(format_parts[2]),
+            animated=format_parts[3] is not None,
+            extension=format_parts[4],
+        )
+
+
+@dataclass(frozen=True, eq=True)
+class ThumbnailFormat(BaseThumbnailFormat):
+    opts: Optional[str] = ""
+
+
+# Note that this is serialized into a JSONB column in the database,
+# and as such fields cannot be removed without a migration.
+@dataclass(frozen=True, eq=True)
+class StoredThumbnailFormat(BaseThumbnailFormat):
+    content_type: str
+    width: int
+    height: int
+    byte_size: int
+
+
+# Formats that we generate; the first animated and non-animated
+# options on this list are the ones which are written into
+# rendered_content.
+THUMBNAIL_OUTPUT_FORMATS = [
+    # For now, we generate relatively large default "thumbnails", so
+    # that clients that do not understand the thumbnailing protocol
+    # (e.g. mobile) get something which does not look pixelated.
+    ThumbnailFormat("webp", 840, 560, animated=True),
+    ThumbnailFormat("webp", 840, 560, animated=False),
+    # 150x100 is the size that the web client uses
+    ThumbnailFormat("webp", 150, 100, animated=True),
+    ThumbnailFormat("webp", 150, 100, animated=False),
+    ThumbnailFormat("jpg", 150, 100, animated=False),
```

Comment:
Because retina monitors scale at 2-pixels-per-logical-pixel, I wonder if the correct resolution to use for the web app's 150x100 display size is actually 300x200?

Reply:
Switched this in #30868, which I'll rebase this onto once it's a bit further along.